### PR TITLE
App instances

### LINKED ISSUE #1118: Add optional `--app` argument to `apptainer instance [start|stop]`
### Version of Apptainer

What version of Apptainer (or Singularity) are you using? Run

`apptainer --version` (or `singularity --version`).

`apptainer version 1.1.5`

### Expected behavior

`apptainer instance start --app app1 foo.sif foo_app`

What did you expect to see when you do...?

`INFO:   app1 instance started successfully`

### Actual behavior

```Error for command "start": unknown flag: --app```

### PR DESCRIPTION
**Pull request by struktured**
## Description of the Pull Request (PR):

Introduces an optional `--app` argument to the `apptainer instance start` command which will leverage
the `%appstart` section if it exists in the definition file to instantiate a particular app as a service. Eg.

````Console
apptainer instance start --app tail tailer.sif tailer
````

with definition file `tailer.def` containing:

```Yaml
%appstart tail
    tail -f /dev/null
```
will create a service called `tailer` which invokes the app `tail` that simply tails `/dev/null` until terminated.

Stopping the instance is the exact same as stopping any other apptainer instance. The app name is no longer necessary- only the instance name is required. For the above the invocation is

```Console
apptainer instance stop tailer
```
In addition to extending the e2e test, also updated the examples where appropriate to indicate the new functionality.

### This fixes or addresses the following GitHub issues:

 - Fixes #1118 

#### Before submitting a PR, make sure you have done the following:

- [x] Read the [Guidelines for Contributing](https://github.com/apptainer/apptainer/blob/main/CONTRIBUTING.md), and this PR conforms to the stated requirements.
- [x] Added changes to the [CHANGELOG](https://github.com/apptainer/apptainer/blob/main/CHANGELOG.md) if necessary according to the [Contribution Guidelines](https://github.com/apptainer/apptainer/blob/main/CONTRIBUTING.md)
- [x] Added tests to validate this PR, linted with `make check`  and tested this PR locally with a `make test`, and `make testall` if possible (see CONTRIBUTING.md).
  - [x] e2e imageverify test extended to test `%appstart`
- [x] Based this PR against the appropriate branch according to the [Contribution Guidelines](https://github.com/apptainer/apptainer/blob/main/CONTRIBUTING.md)
- [x] Added myself as a contributor to the [Contributors File](https://github.com/apptainer/apptainer/blob/main/CONTRIBUTORS.md)

### CODE REVIEW

**Comment by struktured** [comment]
Some notes before this is ready to merge:

 - I couldn't find documented the appropriate merge strategy- is squashing ok? Or is rebasing required, etc..
 - I am not certain I have sufficient test coverage. I modified the e2e image verify test but I think a pure unit test somewhere might make sense too. Open to guidance on how to add a good test.

**Comment by DrDaveD** [comment]
Overall this contribution looks quite good to me.  We do need you to do `git commit -s` to indicate agreement with the DCO as detailed in CONTRIBUTING.md.

> Some notes before this is ready to merge:
> 
>     * I couldn't find documented the appropriate merge strategy- is squashing ok? Or is rebasing required, etc..

I prefer a single commit for things that are all related.  You might want to have a separate commit for removing the extra blanks you saw.  Each commit needs the `-s` option.  I would use `git reset HEAD^1` to back out each commit and then do `git push -f` to update the PR.

>     * I am not certain I have sufficient test coverage. I modified the e2e image verify test but I think a pure unit test somewhere might make sense too. Open to guidance on how to add a good test.

Just look for precedents for similar features.  If they have unit test coverage, add a similar unit test, otherwise don't bother, and let the e2e test cover it.

**Review by DrDaveD — CHANGES_REQUESTED** [review]
I'm sorry for the delay, I was away.  I approved the test runs and there's an [e2e test failure](https://github.com/apptainer/apptainer/actions/runs/4249539640/jobs/7593753494#step:10:3917) and [another set of 3](https://github.com/apptainer/apptainer/actions/runs/4249539640/jobs/7593753494#step:10:4040) just like the first.

**Comment by DrDaveD** [comment]
@struktured can you please look at the e2e test failures?

**Comment by struktured** [comment]
Checking...

**Comment by DrDaveD** [comment]
Still seeing [similar errors](https://github.com/apptainer/apptainer/actions/runs/4412070882/jobs/7744616921#step:10:599) after the latest commits.

**Comment by struktured** [comment]
I got throttled by docker when testing yesterday so decided to push it
hoping it was fixed. Will continue debugging it!

**Comment by struktured** [comment]
@DrDaveD - I can clean up the commit history but before I do that- is there anything else blocking this PR? Thanks.

**Review by DrDaveD — APPROVED** [review]

**Comment by DrDaveD** [comment]
It looks good to me now.  After you clean up the commit history, I will merge.

**Comment by DrDaveD** [comment]
Uh oh, this time the lint markdown failed.  "Headings should be surrounded by blank lines".  It doesn't like the new separater you added before "Other changes".

**Comment by struktured** [comment]
Noted, will fix it after work today.

**Comment by DrDaveD** [comment]
Thanks!  Please also look through the [documentation](https://apptainer.org/docs/user/main) and if changes are needed to reflect the new functionality please submit a PR at https://github.com/apptainer/apptainer-userdocs.  I expect some changes are needed to the definition file page and maybe a brief mention on the instances/running_services page.

**Comment by struktured** [comment]
> Thanks! Please also look through the [documentation](https://apptainer.org/docs/user/main) and if changes are needed to reflect the new functionality please submit a PR at https://github.com/apptainer/apptainer-userdocs. I expect some changes are needed to the definition file page and maybe a brief mention on the instances/running_services page.

Will do so this weekend!

**Pull request merged**
